### PR TITLE
*: add a String method for Datum to make the log more friendly

### PR DESCRIPTION
**Pull request by jackysp**
### What problem does this PR solve? 
Print more info when some errors occur in some DMLs.
1. add a String method for Datum
1. add a log for `replace` and `insert on duplicate update`

### What is changed and how it works?
The error message will look like,
```
[replace] can not find record which handle is 1 for the to-be-inserted row [(Kind: bigint, Value: 1) (Kind: bigint, Value: 1)]
```

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

PTAL @coocood @shenli

### CODE REVIEW

File: executor/insert.go [hunk 1]
```diff
@@ -205,6 +206,7 @@ func (e *InsertExec) Open(ctx context.Context) error {
 func (e *InsertExec) updateDupRow(row toBeCheckedRow, handle int64, onDuplicate []*expression.Assignment) error {
 	oldRow, err := e.getOldRow(e.ctx, e.Table, handle)
 	if err != nil {
+		log.Errorf("[insert on dup] can not find record which handle is %d for the to-be-inserted row %v", handle, row.row)
```
Comment (shenli) [inline executor/insert.go]: Why not use %s?

Comment (jackysp) [reply executor/insert.go]: The lint tool does not allow %s.

File: types/datum.go [hunk 2]
```diff
@@ -1871,3 +1871,38 @@ func CopyRow(dr []Datum) []Datum {
...
+	case KindBytes:
+		return fmt.Sprintf(preStr+"%x)", d.GetBytes())
+	case KindMysqlDecimal:
+		return fmt.Sprintf(preStr+"%s)", d.GetMysqlDecimal())
+	case KindMysqlDuration:
+		return fmt.Sprintf(preStr+"%s)", d.GetMysqlDuration())
+	case KindMysqlEnum:
+		return fmt.Sprintf(preStr+"%s)", d.GetMysqlEnum())
+	case KindBinaryLiteral, KindMysqlBit:
+		return fmt.Sprintf(preStr+"%x)", d.GetBinaryLiteral())
```
Comment (zimulala) [inline types/datum.go]: Could we use `%s`？

Comment (jackysp) [reply types/datum.go]: I prefer to use a consistent format for the byte slice.

File: types/datum.go [hunk 3]
```diff
@@ -1871,3 +1871,38 @@ func CopyRow(dr []Datum) []Datum {
 	}
 	return c
 }
+
+// String implements fmt.Stringer interface.
+func (d Datum) String() string {
```
Comment (lamxTyler) [inline types/datum.go]: Why not implement it based on `Datum.ToString`?

File: executor/insert.go (hunk 1 above)
Comment (CaitinChen) [reply executor/insert.go]: log.Errorf("[insert on dup] cannot find the record whose handle is %d for the to-be-inserted row %v", handle, row.row)

File: executor/replace.go [hunk 4]
```diff
@@ -52,6 +53,7 @@ func (e *ReplaceExec) removeRow(handle int64, r toBeCheckedRow) (bool, error) {
 	newRow := r.row
 	oldRow, err := e.batchChecker.getOldRow(e.ctx, r.t, handle)
 	if err != nil {
+		log.Errorf("[replace] can not find record which handle is %d for the to-be-inserted row %v", handle, r.row)
```
Comment (zhexuany) [inline executor/replace.go]: s/which/whose

File: executor/write_test.go [hunk 5]
```diff
@@ -1995,3 +1998,35 @@ func (s *testSuite) TestUpdateAffectRowCnt(c *C) {
...
+	txn, err := s.store.Begin()
+	c.Assert(err, IsNil)
+	_, err = indexOpr.Create(s.ctx, txn, types.MakeDatums(1), 1)
+	c.Assert(err, IsNil)
+	err = txn.Commit(context.Background())
+	c.Assert(err, IsNil)
+
+	_, err = tk.Exec(`replace into testLog values (0, 0), (1, 1);`)
+	c.Assert(err, NotNil)
+	expErr := errors.New(`can not be duplicated row, due to old row not found. handle 1 not found`)
```
Comment (zhexuany) [inline executor/write_test.go]: If this err can be reused somehow, I prefer you could make it into a const.

Comment (zhexuany) [reply executor/write_test.go]: gotcha.

**Pull request merged**